bjoern/request.c

```diff
@@ -121,50 +121,70 @@ on_path(http_parser* parser, const char* path, size_t len)
 {
   if(!(len = unquote_url_inplace((char*)path, len)))
     return 1;
-  _set_header_free_value(_PATH_INFO, _Unicode_FromStringAndSize(path, len));
+  _set_or_append_header(_PATH_INFO, path, len);
   return 0;
 }
 
 static int
 on_query_string(http_parser* parser, const char* query, size_t len)
 {
-  _set_header_free_value(_QUERY_STRING, _Unicode_FromStringAndSize(query, len));
+  _set_or_append_header(_QUERY_STRING, query, len);
   return 0;
 }
 
 static int
 on_header_field(http_parser* parser, const char* field, size_t len)
 {
-  if(PARSER->value.data) {
-    /* Store previous header and start a new one */
-    _set_header_from_http_header();
-  } else if(PARSER->field.data) {
-    UPDATE_LENGTH(field);
+  if(PARSER->last_was_data) {
+    _free_and_unset_if_set(PARSER->field);
+    PARSER->field = _Unicode_FromStringAndSize("HTTP_", 5);
+    PARSER->last_was_data = false;
+    PARSER->invalid_header = false;
+  }
+
+  if(PARSER->invalid_header) {
     return 0;
   }
-  PARSER->field = (string){(char*)field, len};
-  PARSER->value = (string){NULL, 0};
+
+  char field_processed[len];
```

Comment:
Why not keep the old code that first buffers everything and then calls `PyUnicode_FromString` a single team at the very end? That's easier to understand IMO

Reply:
I'm always appending to PyUnicode, which I am concating. I use this in all callbacks, so i didn't want to make another logic for `char *` buffer with reallocation and so on. And as this I needed to move wsgi converter logic here, which is now processing the input buffer-like, as callbacks works too.
So it can be converted to store string (char *, size_t) instead of PyObject, and then put back the wsgi function as it was, at the moment i see it as a preference on, what should be easier to understand, the convertor, or another new memory management over buffered string.

I can do both, which do you prefer?

Reply:
Let's keep it this way. Since we're using unicode concatenation for header values as well, which makes sense from a simplicity and performance point of view.